pkg/controller/controller_ref_manager.go

```diff
@@ -342,3 +365,20 @@ func (m *ReplicaSetControllerRefManager) ReleaseReplicaSet(replicaSet *extension
 	}
 	return err
 }
+
+// RecheckDeletionTimestamp returns a canAdopt() function to recheck deletion.
+//
+// The canAdopt() function calls getObject() to fetch the latest value,
+// and denies adoption attempts if that object has a non-nil DeletionTimestamp.
+func RecheckDeletionTimestamp(getObject func() (metav1.Object, error)) func() error {
+	return func() error {
+		obj, err := getObject()
+		if err != nil {
+			return fmt.Errorf("can't recheck DeletionTimestamp: %v", err)
+		}
+		if obj.GetDeletionTimestamp() != nil {
+			return fmt.Errorf("%v/%v has just been deleted at %v", obj.GetNamespace(), obj.GetName(), obj.GetDeletionTimestamp())
```

Comment:
I use "is being deleted" in this case.

Users do need to distinguish if an object is "being deleted" or "has been removed from the storage". For example, in the former case, the name is held and a user cannot create an object of the same name. Also, in the context of deleting a pod, "deleted" indicates the resources held by the pod are released, while "being deleted" means kubelet is still reclaiming the resources.